kore/src/Kore/Step/Simplification/OverloadSimplifier.hs

```diff
@@ -0,0 +1,140 @@
+License     : NCSA
+
+-}
+module Kore.Step.Simplification.OverloadSimplifier
+    ( OverloadSimplifier (..)
+    , mkOverloadSimplifier
+    ) where
+
+import Control.Exception
+    ( assert
+    )
+import Data.Function
+    ( on
+    )
+import qualified Data.Map.Strict as Map
+import Data.Maybe
+    ( listToMaybe
+    , mapMaybe
+    )
+import qualified Data.Set as Set
+import qualified GHC.Stack as GHC
+
+import qualified Kore.Attribute.Axiom as Attribute
+import qualified Kore.Attribute.Symbol as Attribute
+import Kore.Attribute.Synthetic
+    ( synthesize
+    )
+import Kore.IndexedModule.IndexedModule
+    ( VerifiedModule
+    , recursiveIndexedModuleAxioms
+    )
+import Kore.IndexedModule.MetadataTools as MetadataTools
+import Kore.IndexedModule.SortGraph
+    ( SortGraph
+    , isSubsortOf
+    )
+import Kore.Internal.ApplicationSorts
+    ( applicationSortsResult
+    )
+import qualified Kore.Internal.Symbol as Symbol
+import Kore.Internal.TermLike
+
+data OverloadSimplifier =
+    OverloadSimplifier
+        { isOverloading :: Symbol -> Symbol -> Bool
+        -- ^ whether the first argument is overloading the second
+        , isOverloaded :: Symbol -> Bool
+        -- ^ Whether the symbol is overloaded
+        , resolveOverloading
+            :: forall variable
+            .  GHC.HasCallStack
+            => InternalVariable variable
+            => Inj ()
+            -> Symbol
+            -> [TermLike variable]
+            -> TermLike variable
+        , unifyOverloadWithinBound :: Symbol -> Symbol -> Sort -> Maybe Symbol
+        }
+
+mkOverloadSimplifier
+    :: VerifiedModule Attribute.Symbol Attribute.Axiom
+    -> SortGraph
+    -> SmtMetadataTools Attribute.StepperAttributes
+    -> OverloadSimplifier
+mkOverloadSimplifier verifiedModule sortGraph tools =
+    OverloadSimplifier
+        { isOverloading
+        , isOverloaded
+        , resolveOverloading
+        , unifyOverloadWithinBound
+        }
+  where
+    isOverloading = checkOverloading `on` Symbol.toSymbolOrAlias
+    isOverloaded = checkOverloaded . Symbol.toSymbolOrAlias
+
+    resolveOverloading injProto overloadedHead overloadingChildren =
+        mkApplySymbol overloadedHead
+        $ assert (length overloadedChildrenSorts == length overloadingChildren)
+        $ zipWith mkInj overloadingChildren overloadedChildrenSorts
+      where
+        overloadedChildrenSorts =
+            Symbol.applicationSortsOperands (symbolSorts overloadedHead)
+        mkInj injChild injTo =
+            (synthesize . InjF) injProto { injFrom, injTo, injChild }
+          where
+            injFrom = termLikeSort injChild
+
+    checkOverloaded :: SymbolOrAlias -> Bool
+    checkOverloaded = (`Set.member` allOverloadsSet)
+
+    checkOverloading :: SymbolOrAlias -> SymbolOrAlias -> Bool
+    checkOverloading head1 head2 = (head1, head2) `Set.member` overloadPairsSet
+
+    overloadPairsSet = Set.fromList overloadPairsList
+
+    overloadPairsList =
```

Comment:
I think we could avoid the back-and-forth between `Internal.Symbol` and `SymbolOrAlias` if we ran the `Overload` attributes through the symbol verifier.